make: Further improvements to feature enable/disable in configure.py.

bradleysepos opened the pull request:
Converts most of configure.py from optparse (deprecated) to argparse. Better fixes #1459.

**Test on:**

- [x] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [x] Ubuntu Linux
- [x] Python 2.7.10
- [x] Python 3.7.2

bradleysepos committed:
make: Further improvements to feature enable/disable in configure.py.

Converts most of configure.py from optparse (deprecated) to argparse. Better fixes #1459.

bradleysepos commented:
Building on 7b035348b714f945cff1ae2940332960a8eb0f00, which disabled features incompatible with the target platform.

This PR is more involved, both properly setting all defaults and hiding help options where desired.

jstebbins commented:
```configure --debug=max``` fails
~~~
$ ../configure --force --debug=max --optimize=none
probe: host tuple...(pass) x86_64-pc-linux-gnu
compute: available architectures...(pass) x86_64
find: ar...(pass) /usr/bin/ar
find: cp...(pass) /usr/bin/cp
find: gcc...(pass) /usr/bin/gcc
find: gmake...(pass) /usr/bin/gmake
find: gm4...(pass) /usr/bin/m4
find: mkdir...(pass) /usr/bin/mkdir
find: gpatch...(pass) /usr/bin/patch
find: rm...(pass) /usr/bin/rm
find: ranlib...(pass) /usr/bin/ranlib
find: strip...(pass) /usr/bin/strip
find: gtar...(pass) /usr/bin/gtar
find: nasm...(pass) /usr/bin/nasm
find: autoconf...(pass) /usr/bin/autoconf
find: automake...(pass) /usr/bin/automake
find: cmake...(pass) /usr/bin/cmake
find: libtool...(pass) /usr/bin/libtool
find: pkg-config...(pass) /usr/bin/pkg-config
find: xcodebuild...(fail) not found
find: lipo...(fail) not found
find: python3...(pass) /usr/bin/python3
usage: configure.py [OPTIONS...] [TARGETS...]
configure.py: error: argument --debug: ignored explicit argument 'max'
~~~

bradleysepos commented:
Probably something with `add_argument` and `nargs` param.

I probably won't be able to look at it speedily on account of illness.

bradleysepos commented:
See https://docs.python.org/3/library/argparse.html and https://mkaz.blog/code/python-argparse-cookbook/

Probably needs `nargs='?'` if the value is optional, or 1 if required.

jstebbins commented:
Also need to change ```'store_const'``` to ```'store'``` since store_const does not accept nargs argument.
Something like:
~~~
diff --git a/make/configure.py b/make/configure.py
index 95c4c1b67..725abba48 100644
--- a/make/configure.py
+++ b/make/configure.py
@@ -655,9 +655,15 @@ class SelectMode( dict ):
         self.mode = self.default
 
     def cli_add_argument( self, parser, option ):
-        parser.add_argument( option, default=self.mode, metavar='MODE',
+        if self.modes:
+            if self.default:
+                nargs_val = '?'
+            else:
+                nargs_val = 1
+
+        parser.add_argument( option, nargs=nargs_val, default=self.mode, metavar='MODE',
             help='select %s%s: %s' % (self.descr,self.what,self.toString()),
-            action='store_const', const=lambda:'self.cli_callback' )
+            action='store', const=lambda:'self.cli_callback')
 
     def cli_callback( self, option, opt_str, value, parser, *args, **kwargs ):
         if value not in self:
~~~
This is all pretty much magic to me.

bradleysepos commented:
Does the lambda execute with store? If so, works for me.

jstebbins committed:
configure: fix parsing debug and optimize options

required and optional arguments to an option requires setting nargs in
add_argument

jstebbins commented:
I don't know what the lambda is doing.  Like is said, it's FM

jstebbins commented:
Nope, something's not right. Now it's just setting defaults and ignoring supplied argument.

bradleysepos commented:
Should be calling the function cli_callback right below that block. I might have to bug one of our resident Python people who have helped before, argparse is pretty obtuse to me. Amazed I got that far.

jstebbins commented:
I'm reading argparse.py.  It looks like StoreAction does nothing with const.  So something else will need to be done.

bradleysepos commented:
@qarkai Any chance you can help with this?

qarkai commented:
@bradleysepos I'll take a look on the weekend if it's not urgent.

bradleysepos commented:
Not urgent, thank you!

jstebbins commented:
I don't really understand the lambda syntax you used.  I can't find any reference that looks at all similar.  I made it work though by changing the lambda and adding a custom action.  e.g.
~~~
diff --git a/make/configure.py b/make/configure.py
index 9aa4883fe..069d67768 100644
--- a/make/configure.py
+++ b/make/configure.py
@@ -642,6 +642,13 @@ class CoreProbe( Action ):
 
 ###############################################################################
 
+class HBStoreConstAction(argparse.Action):
+    def __init__(self, option_strings, dest, nargs=None, **kwargs):
+        super(HBStoreConstAction, self).__init__(
+            option_strings, dest, nargs, **kwargs)
+    def __call__(self, parser, namespace, values, option_string=None):
+        setattr(namespace, self.dest, self.const(values))
+
 class SelectMode( dict ):
     def __init__( self, descr, *modes, **kwargs ):
         super( SelectMode, self ).__init__( modes )
@@ -657,9 +664,9 @@ class SelectMode( dict ):
     def cli_add_argument( self, parser, option ):
         parser.add_argument( option, nargs='?', default=self.mode, metavar='MODE',
             help='select %s%s: %s' % (self.descr,self.what,self.toString()),
-            action='store', const=lambda:'self.cli_callback')
+            action=HBStoreConstAction, const=lambda x:self.cli_callback(x))
 
-    def cli_callback( self, option, opt_str, value, parser, *args, **kwargs ):
+    def cli_callback( self, value ):
         if value not in self:
             raise ArgumentError( 'invalid %s%s: %s (choose from: %s)'
                 % (self.descr,self.what,value,self.toString( True )) )
~~~

bradleysepos commented:
Seems like a good solution to me. I'm not sure we need to HB* namespace this, though.

jstebbins commented:
Right HB prefix isn't needed.

Do you have a link or something that illustrates how you are using lambda?  I would like to understand how the other occurrences of lambda in configure.py are working and verify they are correct.

bradleysepos commented:
Either StackOverflow'd it or fat fingered it. Assume the others are wrong as well. Your solution seems clean.

jstebbins commented:
My previous solution failed when the optional argument was not supplied, i.e. ```--optimize=speed``` works but ```--optimize``` does not.  But I think I have found the *Right*(TM) solution.  Try this on for size.
~~~
diff --git a/make/configure.py b/make/configure.py
index 95c4c1b67..f57f54516 100644
--- a/make/configure.py
+++ b/make/configure.py
@@ -642,6 +642,18 @@ class CoreProbe( Action ):
 
 ###############################################################################
 
+class StoreCallbackAction(argparse.Action):
+    def __init__(self, option_strings, dest, nargs=None, **kwargs):
+        self.callback = kwargs.pop('callback', None)
+
+        super(StoreCallbackAction, self).__init__(
+            option_strings, dest, nargs, **kwargs)
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        setattr(namespace, self.dest, values)
+        if self.callback != None:
+            self.callback(self, values)
+
 class SelectMode( dict ):
     def __init__( self, descr, *modes, **kwargs ):
         super( SelectMode, self ).__init__( modes )
@@ -655,14 +667,16 @@ class SelectMode( dict ):
         self.mode = self.default
 
     def cli_add_argument( self, parser, option ):
-        parser.add_argument( option, default=self.mode, metavar='MODE',
+        parser.add_argument(option, nargs='?', metavar='MODE',
+            default=self.mode, const=self.mode,
             help='select %s%s: %s' % (self.descr,self.what,self.toString()),
-            action='store_const', const=lambda:'self.cli_callback' )
+            action=StoreCallbackAction, callback=self.cli_callback)
 
-    def cli_callback( self, option, opt_str, value, parser, *args, **kwargs ):
+    def cli_callback( self, action, value ):
         if value not in self:
-            raise ArgumentError( 'invalid %s%s: %s (choose from: %s)'
-                % (self.descr,self.what,value,self.toString( True )) )
+            raise argparse.ArgumentError(action,
+                'invalid %s%s: %s (choose from: %s)'
+                % (self.descr, self.what, value, self.toString(True)))
         self.mode = value
 
     def toString( self, nodefault=False ):
~~~

bradleysepos commented:
At the bottom of https://docs.python.org/2/library/argparse.html, upgrading from optparse to argparse:

> Replace callback actions and the callback_* keyword arguments with type or action arguments.

IIRC the callback param isn't a thing in ~Python 3~ argparse, which is why I tried to use lambda functions as a workaround. Reimplementing it is clever, though.

bradleysepos commented:
I wonder whether simply, `type=lambda x:self.cli_callback(x)` does what we want without reimplementing the callback syntax?

jstebbins commented:
Right.  And there are some stackoverflow suggestions to use lambda in various ways with argparse, but they are all abusing internal behaviour in some way that could be broken by future changes in the python library.  

I thought of using type for that, but type is meant to be used to convert a string to the target type and could conceivably be used by the library to do conversion for some reason when *not* setting the value.  So it didn't seem prudent to abuse it that way.  The problem with using const turned out to be that the value of const is what is returned when you leave off an optional parameter.  So the value passed to the callback in that case turned out to be the reference to the callback :stuck_out_tongue: 

I think what I've done is fairly immune to potential changes in the python library and is a pretty versatile pattern to use for this.

bradleysepos commented:
Works for me. When in doubt, read the source! Thanks, John.

jstebbins commented:
The SelectTool class does not appear to actually be initialized anywhere.  I.e. SelectTool.selects[] is always empty and never gets set.  Any idea what that's all about?  What is SelectTool anyway?

jstebbins commented:
Ah, I see what it's meant for.  When there are multiple options for a tool (e.g. wget, curl), it selects one.  It is no longer in use because we eliminated wget and curl.

jstebbins committed:
configure: fix argparse callbacks

argparse does not have callbacks like optparse did, but a custom action
can be used to perform the callback.

jstebbins commented:
I've tested my changes with python 2 & 3.  Should probably smoke test the other platforms again, then I think it is good to go.

bradleysepos commented:
To really make this complete, we could replace the remaining optparse code with argparse: https://github.com/HandBrake/HandBrake/blob/dd6ed2996cbc2e6b6a14b9cd8d69dcdd0d6e147e/make/configure.py#L1245-L1272

bradleysepos committed:
make: Fix configure and Xcode where cross compiling.

bradleysepos committed:
make: Tidy up configure help.

Should no longer need to specify platform in the help, since the help now only shows options available for the specified build.

bradleysepos commented:
Looks like we'll have to replace that final optparse code; it doesn't work anymore. If the build is reconfigured, it probably will be configured without all the specified options. ☹️

bradleysepos committed:
make: Fix configure.py conf args for reconfigure.

bradleysepos committed:
make: Improve handling of cross argument in configure.py.

bradleysepos committed:
make: Rerun cross tools probe before creating CLI help.

bradleysepos committed:
make: Fix configure --minver argument showing for cross.

bradleysepos committed:
make: Remove last optparse import in configure.py.

Now fully transitioned to argparse.

bradleysepos committed:
make: Remove SelectTool class from configure.py.

Has not been in use for a long time.

bradleysepos committed:
make: Fix initialization of minver in configure.py.

bradleysepos committed:
make: Reflect FFmpeg AAC required on some platforms in help.

bradleysepos committed:
make: Tweak configure.py output language.

Internally, configure.py may use the terms host and build somewhat inappropriately, but that does not mean we should print such ambiguity.

bradleysepos commented:
Should be ready for final review. If no concerns, I'll merge soon.

bradleysepos commented:
I went ahead and tested Linux so you wouldn't have to again. 😸
